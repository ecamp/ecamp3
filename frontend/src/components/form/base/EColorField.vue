<template>
  <EParseField
    ref="input"
    :value="value"
    :format="format"
    :parse="parse"
    :serialize="serialize"
    :deserialize="deserialize"
    :required="required"
    :vee-id="veeId"
    :vee-rules="veeRules"
    reset-on-blur
    v-bind="$attrs"
    v-on="$listeners"
    @input="$emit('input', $event)"
  >
    <!-- passing through all slots -->
    <slot v-for="(_, name) in $slots" :slot="name" :name="name" />
    <template v-for="(_, name) in $scopedSlots" :slot="name" slot-scope="slotData">
      <slot v-if="name !== 'prepend'" :name="name" v-bind="slotData" />
    </template>
    <template #prepend="props">
      <slot name="prepend" v-bind="props">
        <ColorSwatch
          class="mt-n1"
          :color="props.serializedValue"
          tag="div"
          :aria-label="props.serializedValue"
        />
      </slot>
    </template>
  </EParseField>
</template>

<script>
import { reactive } from 'vue'
import { formComponentMixin } from '@/mixins/formComponentMixin.js'
import { parse, serialize } from 'colorjs.io/fn'
import ColorSwatch from '@/components/form/base/ColorPicker/ColorSwatch.vue'

export default {
  name: 'EColorField',
  components: { ColorSwatch },
  mixins: [formComponentMixin],
  props: {
    value: { type: String, required: false, default: null },
  },
  emits: ['input'],
  methods: {
    format(value) {
      if (typeof value === 'string') {
        return value
      }
      return !value
        ? ''
        : serialize(value, {
            space: 'srgb',
            format: 'hex',
            collapse: false,
          }).toUpperCase()
    },
    /**
     * @param {string} value
     */
    parse(value) {
      if (value === '') {
        return null
      }
      try {
        const color = parse(value)
        color.alpha = 1
        return reactive(color)
      } catch (e) {
        if (e instanceof TypeError) {
          throw new Error(this.$tc('components.form.base.eColorField.parseError'))
        } else {
          throw e
        }
      }
    },
    /**
     * @param {string|null} value
     */
    serialize(value) {
      try {
        return serialize(value, { format: 'hex', collapse: false }).toUpperCase()
      } catch (e) {
        return null
      }
    },
    /**
     * @param value {null|string}
     * @return {null|Color}
     */
    deserialize(value) {
      try {
        return !value ? null : reactive(parse(value, { space: 'srgb', format: 'hex' }))
      } catch (e) {
        return null
      }
    },
    focus() {
      this.$refs.input.focus()
    },
  },
}
</script>
