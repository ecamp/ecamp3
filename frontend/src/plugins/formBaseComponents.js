// Global registration of all Vue components in folder components/form/base

import ECheckbox from '@/components/form/base/ECheckbox.vue'
import EColorPicker from '@/components/form/base/EColorPicker.vue'
import EDatePicker from '@/components/form/base/EDatePicker.vue'
import EForm from '@/components/form/base/EForm.vue'
import ERichtext from '@/components/form/base/ERichtext.vue'
import ESelect from '@/components/form/base/ESelect.vue'
import ESwitch from '@/components/form/base/ESwitch.vue'
import ETextarea from '@/components/form/base/ETextarea.vue'
import ETextField from '@/components/form/base/ETextField.vue'
import ETimePicker from '@/components/form/base/ETimePicker.vue'

class FormBaseComponentsPlugin {
  install(Vue) {
    Vue.component('ECheckbox', ECheckbox)
    Vue.component('EColorPicker', EColorPicker)
    Vue.component('EDatePicker', EDatePicker)
    Vue.component('EForm', EForm)
    Vue.component('ERichtext', ERichtext)
    Vue.component('ESelect', ESelect)
    Vue.component('ESwitch', ESwitch)
    Vue.component('ETextarea', ETextarea)
    Vue.component('ETextField', ETextField)
    Vue.component('ETimePicker', ETimePicker)
  }
}

export default new FormBaseComponentsPlugin()
