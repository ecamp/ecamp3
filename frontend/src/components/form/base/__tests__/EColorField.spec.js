import Vue from 'vue'
import Vuetify from 'vuetify'

import formBaseComponents from '@/plugins/formBaseComponents'

import EColorField from '@/components/form/base/EColorField.vue'
import { mount as mountComponent } from '@vue/test-utils'
import { ColorSpace, sRGB } from 'colorjs.io/fn'

Vue.use(Vuetify)
Vue.use(formBaseComponents)

ColorSpace.register(sRGB)

describe('An EColorField', () => {
  let vuetify

  const mount = (options) => {
    const app = Vue.component('App', {
      components: { EColorField },
      data: function () {
        return {
          data: null,
        }
      },
      methods: {
        parse: (value) => {
          return value === 'true' ? true : value === 'false' ? false : null
        },
        format: (value) => {
          return value === null ? '' : `${value}`
        },
      },
      template: `<div data-app><e-color-field label="test" v-model="data"/></div>`,
    })
    return mountComponent(app, { vuetify, attachTo: document.body, ...options })
  }

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  test('looks like a textfield', async () => {
    const wrapper = mount()
    expect(wrapper).toMatchSnapshot('empty')

    await wrapper.setData({ data: '#FF0000' })
    expect(wrapper).toMatchSnapshot('with text')
  })

  test('updates text when vModel changes', async () => {
    const wrapper = mount()
    const input = wrapper.find('input').element
    expect(input.value).toBeDefined()

    const firstValue = '#A1E1E1'
    await wrapper.setData({ data: firstValue })
    expect(input.value).toBe(`${firstValue}`)

    const secondValue = '#FF00FF'
    await wrapper.setData({ data: secondValue })
    expect(input.value).toBe(`${secondValue}`)
  })

  test('updates vModel when value of input field changes', async () => {
    const wrapper = mount()
    const input = wrapper.find('input')
    const value = '#123456'

    input.element.value = `${value}`
    await input.trigger('input')

    expect(wrapper.vm.data).toBe(value)
  })

  test.each([
    ['#00FF00', '#00FF00'],
    ['red', '#FF0000'],
    ['', null],
  ])('parses "%s" as "%s"', async (string, expected) => {
    const wrapper = mount()
    const input = wrapper.find('input')

    input.element.value = `${string}`
    await input.trigger('input')

    expect(wrapper.vm.data).toBe(expected)
  })
})
