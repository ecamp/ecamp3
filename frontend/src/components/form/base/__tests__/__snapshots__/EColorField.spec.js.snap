// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`An EColorField > looks like a textfield > empty 1`] = `
<div data-app>
     <div class="e-form-container"
          label="test">
          <div class="v-input v-input--hide-details theme--light v-text-field v-text-field--filled v-text-field--enclosed">
               <div class="v-input__prepend-outer">
                    <div class="e-colorswatch mt-n1 v-btn v-btn--fab v-btn--has-bg v-btn--round theme--light elevation-0 v-size--default e-colorswatch--null"
                         style="height: 30px; width: 30px; --ee5e0d96-contrast: black;"
                         tag="div">
                         <span class="v-btn__content"></span>
                    </div>
               </div>
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-text-field__slot">
                              <label class="v-label theme--light"
                                     for
                                     style="left: 0px; position: absolute;">test</label>
                              <input id
                                     type="text">
                         </div>
                    </div>
               </div>
          </div>
     </div>
</div>
`;

exports[`An EColorField > looks like a textfield > with text 1`] = `
<div data-app>
     <div class="e-form-container"
          label="test">
          <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--filled v-text-field--enclosed">
               <div class="v-input__prepend-outer">
                    <div aria-label="#FF0000"
                         class="e-colorswatch mt-n1 v-btn v-btn--fab v-btn--has-bg v-btn--round theme--light elevation-0 v-size--default"
                         style="height: 30px; width: 30px; --ee5e0d96-contrast: #fff; background-color: rgb(255, 0, 0); border-color: #ff0000;"
                         tag="div">
                         <span class="v-btn__content"></span>
                    </div>
               </div>
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-text-field__slot">
                              <label class="v-label v-label--active theme--light"
                                     for
                                     style="left: 0px; position: absolute;">test</label>
                              <input id
                                     type="text">
                         </div>
                    </div>
               </div>
          </div>
     </div>
</div>
`;
